Php8 mail migration

fhelfer opened the pull request:
PHP 8 Migration: 
* Return-Type-Declarations: function foo() : void
* Type-Hints: function(string $a) : void
* Type Properties: private ilObjUser $user;
* `declare(strict_types=1);`

in:
Services/Mail
Services/Contact/classes

fhelfer committed:
Services/Mail & Services/Contact/classes Code Smell Fixes

contacts code smell fixing

substr replace, const declare, strlen replace Contact/classes

Contact/classes array_push changed to arr[]= for performance

fhelfer committed:
Services/Mail & Services/Contact/classes added strict types, return types declarations, type hints, type properties

Contact/classes added strict types #2

Contact/classes added strict types #3

Mail/Address added strict types

Contact/classes Type Properties

mail added strict types

Services/Mail/classes addded strict types

Services/Mail/classes added strict types + Services/Contact/ template type fix

Services/Mail/classes added strict types 2

Services/Mail/classes added strict types 3

Services/Mail/classes added strict types 4

Services/Mail/classes added strict types 5

Services/Mail/classes added strict types 6

reverted changes to switch statements

revert unwanted changes

bugifx mixed default typ

mjansenDatabay reviewed (changes requested):
Hi @fhelfer , thanks for this PR.

Please try to solve the following issues:

1. This PR contains files NOT located in `Services/Mail` or `Services/Contact/classes`. Please remove these files from the change list (`ilCtrl`, `ilTable2GUI`, `php_errors.log`).
2. Please try to `squash` the commits.
3. If you change a file, please try to apply the [`PSR-2 + X` code style](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/development/coding-style.md).
4. I still find many usages of `$_POST` or `$_GET`. Please try to replace them by using the corresponding methods of `$DIC->http()->request()`. Where applicable, please use our `Refinery` transformations to ensure the user input has the correct type/shape when being processed (example: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/Modules/Chatroom/classes/gui/class.ilChatroomPrivateRoomGUI.php#L72).
5. Try to replace read/write operations for `$_SESSION` with `ilSession::get` and `ilSession:set` .

Important: Try to execute the code by performing the necessary user interface operations in the mail system (e.g. _Send a mail_, _Add an attachment_, etc.).

If you have any questions or need help, don't hesitate to contact me.

Best regards,
@mjansenDatabay

fhelfer committed:
psr2 code style fixer

fhelfer committed:
Services/Mail & Services/Contact/classes bugfixing

fixed most errors encountered

mail fixes

fhelfer committed:
replaced GET/POST direct access from Services/Mail & Services/Contact/classes

replace  in Services/Mail

replace  &  in Services/Mail

GET,POST,SESSION replaced in Services/Contact/classes

bugfixing, moved GET manipulation to ilSession

moved GET/POST manipulation to ilSession, bugfix

fhelfer committed:
removed GET/POST manipulation in Services/Mail & Services/Contact/classes

GET Manipulation fix view

GET/POST manipulation fixes

revert changes

remove unwanted files

removed unwanted files

bugfix

fhelfer committed:
PSR-2 cody style fix in Servces/Mail & Services/Contact/classes

fhelfer committed:
removed REQUEST from Services/Contact/classes

mjansenDatabay commented:
@fhelfer Could you please resolve the conflicts?

fhelfer commented:
> @fhelfer Could you please resolve the conflicts?

@mjansenDatabay Done

mjansenDatabay commented on Services/Contact/classes/class.ilContactGUI.php:
```diff
@@ -14,63 +16,25 @@
-    protected $has_sub_tabs = false;
+    private array|null $postUsrId = null;
+
+    protected ilGlobalPageTemplate $tpl;
```
Please change this to `ilGlobalTemplateInterface`.

mjansenDatabay commented on Services/Contact/classes/class.ilContactGUI.php:
```diff
@@ -14,63 +16,25 @@
-
+    public const CONTACTS_VIEW_GALLERY = 1;
+    public const CONTACTS_VIEW_TABLE = 2;
+    private RequestInterface $httpRequest;
```
IMO `ServerRequestInterface` is more appropriate here.

mjansenDatabay commented on Services/Contact/classes/class.ilContactGUI.php:
```diff
@@ -178,21 +143,21 @@ public function executeCommand()
                 $this->tabs_gui->addSubTab('my_contacts', $this->lng->txt('my_contacts'), $this->ctrl->getLinkTarget($this));
 
-                if (in_array(strtolower($this->ctrl->getCmdClass()), array_map('strtolower', array('ilUsersGalleryGUI', get_class($this))))) {
+                if (in_array(strtolower($this->ctrl->getCmdClass()), array_map('strtolower', array('ilUsersGalleryGUI', get_class($this))), true)) {
```
Please use `ilUsersGalleryGUI::class` instead (here and everywhere class strings are used).

mjansenDatabay commented on Services/Contact/classes/class.ilContactGUI.php:
```diff
@@ -14,63 +16,25 @@
-     * @var bool
-     */
-    protected $has_sub_tabs = false;
+    private array|null $postUsrId = null;
```
ILIAS 8 must be compatible to PHP 7.4 as well. So `Union Types` are not allowed, yet. Here you can use `?array` instead.

mjansenDatabay commented on Services/Contact/classes/class.ilContactGUI.php:
```diff
@@ -14,63 +16,25 @@
+    private RequestInterface $httpRequest;
     /**
-     * @var ilCtrl
+     * @var int[]
```
```suggestion
     * @var int[]|null
```

mjansenDatabay commented on Services/Contact/classes/class.ilContactGUI.php:
```diff
@@ -426,7 +386,7 @@ public function submitInvitation()
         foreach ($usr_ids as $usr_id) {
             $login = ilObjUser::_lookupLogin($usr_id);
-            if (!strlen($login)) {
+            if ($login === '') {
```
`ilObjUser::_lookupLogin` might also return `false` (at the moment).

mjansenDatabay commented on Services/Contact/classes/class.ilContactGUI.php:
```diff
@@ -512,18 +472,32 @@ public function submitInvitation()
-        $usr_ids = $_POST['usr_id'];
+
+        /** @var $DIC \ILIAS\DI\Container */
+        global $DIC;
```
Please move the `global $DIC` to the constructor and initizalize a `$this->refinery` member variable. Don`t use `global` in instance methods.

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSearchGUI.php:
```diff
@@ -1,66 +1,47 @@
+<?php declare(strict_types=1);
 /* Copyright (c) 1998-2015 ILIAS open source, Extended GPL, see docs/LICENSE */
 
+use Psr\Http\Message\RequestInterface;
```
IMO `ServerRequestInterface` is more appropriate here.

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSearchGUI.php:
```diff
@@ -1,66 +1,47 @@
-    protected $mail_roles;
-
+    private RequestInterface $httpRequest;
+    protected mixed $mail_roles;
```
The `mixed` type ist not allowe in PHP 7.4.

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSearchGUI.php:
```diff
@@ -1,66 +1,47 @@
-     * @var ilAbstractMailMemberRoles
-     */
-    protected $objMailMemberRoles;
+    protected mixed $gui;
```
The `mixed` type ist not allowe in PHP 7.4.

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSearchGUI.php:
```diff
@@ -1,66 +1,47 @@
      * @param ilAbstractMailMemberRoles $objMailMemberRoles
      */
-    public function __construct($gui, $ref_id, ilAbstractMailMemberRoles $objMailMemberRoles)
+    public function __construct(mixed $gui, int $ref_id, ilAbstractMailMemberRoles $objMailMemberRoles)
```
The `mixed` type ist not allowe in PHP 7.4.

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSessionRoles.php:
```diff
@@ -54,14 +47,12 @@ public function getMailRoles($ref_id)
-                    $sorted_role_ids[1]['form_option_title'] = $this->lng->txt('send_mail_participants');
-                    break;
-
+            $i = substr($role_title, 0, 12);
```
Please rename this to `$role_prefix` (or something else)?

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchCoursesMembersTableGUI.php:
```diff
@@ -1,6 +1,8 @@
+<?php declare(strict_types=1);
 /* Copyright (c) 1998-2009 ILIAS open source, Extended GPL, see docs/LICENSE */
 
+use Psr\Http\Message\RequestInterface;
```
IMO `ServerRequestInterface` is more appropriate here (and for the other classes where you introduced it).

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchCoursesTableGUI.php:
```diff
@@ -9,11 +9,15 @@
 * @ingroup ServicesMail
 */
+
+use Psr\Http\Message\RequestInterface;
```
IMO `ServerRequestInterface` is more appropriate here (and for the other classes where you introduced it).

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchGUI.php:
```diff
@@ -21,6 +21,9 @@
     +-----------------------------------------------------------------------------+
 */
 
+use JetBrains\PhpStorm\NoReturn;
```
ILIAS does not have a `JetBrains` dependency. Please remove this (here and everywhere).

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchGUI.php:
```diff
@@ -21,6 +21,9 @@
 */
 
+use JetBrains\PhpStorm\NoReturn;
+use Psr\Http\Message\RequestInterface;
```
IMO `ServerRequestInterface` is more appropriate here (and for the other classes where you introduced it).

mjansenDatabay commented on Services/Mail/classes/class.ilPDMailGUI.php:
```diff
@@ -1,6 +1,8 @@
+<?php declare(strict_types=1);
 /* Copyright (c) 1998-2012 ILIAS open source, Extended GPL, see docs/LICENSE */
 
+use Psr\Http\Message\RequestInterface;
```
IMO `ServerRequestInterface` is more appropriate here (and for the other classes where you introduced it).

mjansenDatabay commented on Services/Mail/classes/class.ilPDMailBlockGUI.php:
```diff
@@ -1,6 +1,8 @@
+<?php declare(strict_types=1);
 /* Copyright (c) 1998-2012 ILIAS open source, Extended GPL, see docs/LICENSE */
 
+use Psr\Http\Message\RequestInterface;
```
IMO `ServerRequestInterface` is more appropriate here (and for the other classes where you introduced it).

mjansenDatabay commented on Services/Mail/classes/class.ilObjMailGUI.php:
```diff
@@ -1,22 +1,22 @@
 
-    const PASSWORD_PLACE_HOLDER = '***********************';
+    public const PASSWORD_PLACE_HOLDER = '***********************';
+    private RequestInterface $httpRequest;
```
IMO `ServerRequestInterface` is more appropriate here (and for the other classes where you introduced it).

mjansenDatabay reviewed (changes requested):
- [ ] Please don't use PHP 8 features like `Union Types` or `mixed` (as a return type). ILIAS 8 must be compatible with PHP 7.4 as well.
- [ ] Please used the `static` keyword (if appropriate) and a return type for every function callback (e.g. in `Services/Contact/classes/Provider/ContactMainBarProvider.php`).
- [ ] Please remove obsolete type casts (e.g. `(bool) (ilBuddySystem::getInstance()->isEnabled())`). We don't need them if we already have explicit types.
- [ ] Please replace the expressive array syntax (`array()`) with the short one (`[]`).
- [ ] Please remove all require_once and include_once statements which we do not need anymore. Please keep those in the `tests` folder.
- [ ] Please remove PHPDocs where no added value is provided regarding the PHP types or description.
- [ ] Please add PHPDoc for arrays if they could provide more type related info, e.g. if an array is a list (`int[]`, `string[]`, ...) or a map (`array<int, string>` etc.), or a struct-like array with not more than 4 properties (`array{firstname: string, lastname: string, tite: string, id: int}`). You don't need to apply this for methods derived from base classes of other components (other maintainers).
- [ ] Please use the `<Class>::class` property instead of string, e.g. `ilUsersGalleryGUI::class` instead of `'ilUsersGalleryGUI'`. Class strings are mostly used in `executeCommand` methods, or in method calls like `ilCtrl::getLinkTargetByClass`, `ilCtrl::getFormActionByClass`.

mjansenDatabay commented on Services/Mail/classes/Form/class.ilManualPlaceholderInputGUI.php:
```diff
@@ -9,40 +9,13 @@
+    protected string $dependencyElementId;
+    protected string $instructionText = '';
+    protected string $adviseText = '';
+    protected ilGlobalPageTemplate $tpl;
```
Should be `ilGlobalTemplateInterface`

mjansenDatabay commented on Services/Mail/classes/Mime/Sender/class.ilMailMimeSenderFactory.php:
```diff
@@ -7,14 +7,11 @@
-    /** @var int */
-    protected $anonymousUsrId = 0;
+    protected array $senders = [];
+    protected ?int $anonymousUsrId = 0;
```
`$anonymousUsrId` as a class member should never be `null`, so you can use `int` here.

mjansenDatabay commented on Services/Mail/classes/Mime/Sender/class.ilMailMimeSenderSystem.php:
```diff
@@ -24,7 +23,7 @@ public function __construct(ilSetting $settings)
     public function hasReplyToAddress() : bool
     {
-        return strlen($this->settings->get('mail_system_sys_reply_to_addr')) > 0;
+        return $this->settings->get('mail_system_sys_reply_to_addr') !== '';
```
```suggestion
        return ((string) $this->settings->get('mail_system_sys_reply_to_addr', '')) !== '';
```

mjansenDatabay commented on Services/Mail/classes/Mime/Sender/class.ilMailMimeSenderSystem.php:
```diff
@@ -48,7 +47,7 @@ public function getReplyToName() : string
     public function hasEnvelopFromAddress() : bool
     {
-        return strlen($this->settings->get('mail_system_sys_env_from_addr')) > 0;
+        return $this->settings->get('mail_system_sys_env_from_addr') !== '';
```
```suggestion
        return ((string) $this->settings->get('mail_system_sys_env_from_addr', '')) !== '';
```

mjansenDatabay commented on Services/Mail/classes/Mime/Sender/class.ilMailMimeSenderUser.php:
```diff
@@ -40,7 +37,7 @@ public function getReplyToAddress() : string
             true === (bool) $this->settings->get('use_global_reply_to_addr') &&
             is_string($this->settings->get('global_reply_to_addr', '')) &&
-            strlen($this->settings->get('global_reply_to_addr', '')) > 0
+            $this->settings->get('global_reply_to_addr', '') !== ''
```
```suggestion
            ((string) $this->settings->get('global_reply_to_addr', '')) !== ''
```

mjansenDatabay commented on Services/Mail/classes/Mime/Sender/class.ilMailMimeSenderUser.php:
```diff
@@ -61,7 +58,7 @@ public function getReplyToName() : string
     public function hasEnvelopFromAddress() : bool
     {
-        return strlen($this->settings->get('mail_system_usr_env_from_addr')) > 0;
+        return $this->settings->get('mail_system_usr_env_from_addr') !== '';
```
```suggestion
        return ((string) $this->settings->get('mail_system_usr_env_from_addr', '')) !== '';
```

mjansenDatabay commented on Services/Mail/classes/Mime/Sender/class.ilMailMimeSenderUser.php:
```diff
@@ -86,7 +83,7 @@ public function getFromAddress() : string
     public function getFromName() : string
     {
         $from = $this->settings->get('mail_system_usr_from_name', '');
```
```suggestion
        $from = (string) $this->settings->get('mail_system_usr_from_name', '');
```

mjansenDatabay commented on Services/Mail/classes/Message/class.ilMailBody.php:
```diff
@@ -9,7 +9,7 @@ class ilMailBody
      * @var mixed|null|string|string[]
      */
-    private $bodyContent;
+    private mixed $bodyContent;
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/Mime/Subject/class.ilMailMimeSubjectBuilder.php:
```diff
@@ -52,7 +50,7 @@ public function subject(string $subject, bool $addPrefix = false, string $contex
             }
 
-            if (strlen($prefix) > 0) {
+            if ($prefix !== '') {
```
If we use this check, we have to ensure that the variable is of type `string`. It might be also `null` here.

mjansenDatabay commented on Services/Mail/classes/Mime/Transport/class.ilMailMimeTransportBase.php:
```diff
@@ -148,8 +142,8 @@ final public function send(ilMimeMail $mail) : bool
             if (
-                strpos($message, 'Invalid address') !== false ||
-                strpos($message, 'Message body empty') !== false
+                str_contains($message, 'Invalid address') ||
```
`str_contains` is not supported in PHP 7.4 (https://www.php.net/manual/en/function.str-contains.php), which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/Mime/Transport/class.ilMailMimeTransportBase.php:
```diff
@@ -148,8 +142,8 @@ final public function send(ilMimeMail $mail) : bool
-                strpos($message, 'Invalid address') !== false ||
-                strpos($message, 'Message body empty') !== false
+                str_contains($message, 'Invalid address') ||
+                str_contains($message, 'Message body empty')
```
`str_contains` is not supported in PHP 7.4 (https://www.php.net/manual/en/function.str-contains.php), which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/Mime/Transport/class.ilMailMimeTransportSmtp.php:
```diff
@@ -15,7 +15,7 @@ protected function onBeforeSend() : void
         $this->mailer->Host = $this->settings->get('mail_smtp_host');
         $this->mailer->Port = (int) $this->settings->get('mail_smtp_port');
-        if (strlen($this->settings->get('mail_smtp_user')) > 0) {
+        if ($this->settings->get('mail_smtp_user') !== '') {
```
```suggestion
        if (((string) $this->settings->get('mail_smtp_user', '')) !== '') {
```

mjansenDatabay commented on Services/Mail/classes/class.ilAccountMail.php:
```diff
@@ -303,28 +303,35 @@ public function replacePlaceholders($a_string, $a_user, $a_amail, $a_lang)
+                $this->getUserPassword(),
+                ILIAS_HTTP_PATH . "/login.php?client_id=" . CLIENT_ID,
+                CLIENT_NAME,
+                $ilSetting->get("admin_email"), ],
```
The formatting seems to be not `PSR-2 +x` here.

mjansenDatabay commented on Services/Mail/classes/class.ilMailAutoCompleteSentMailsRecipientsProvider.php:
```diff
@@ -12,53 +12,57 @@ class ilMailAutoCompleteSentMailsRecipientsProvider extends ilMailAutoCompleteRe
             (
-                strpos($this->data['login'], ',') !== false ||
-                strpos($this->data['login'], ';') !== false
+                str_contains($this->data['login'], ',') ||
```
`str_contains` is not supported in PHP 7.4 (https://www.php.net/manual/en/function.str-contains.php), which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/class.ilMailAutoCompleteSentMailsRecipientsProvider.php:
```diff
@@ -12,53 +12,57 @@ class ilMailAutoCompleteSentMailsRecipientsProvider extends ilMailAutoCompleteRe
-                strpos($this->data['login'], ',') !== false ||
-                strpos($this->data['login'], ';') !== false
+                str_contains($this->data['login'], ',') ||
+                str_contains($this->data['login'], ';')
```
`str_contains` is not supported in PHP 7.4 (https://www.php.net/manual/en/function.str-contains.php), which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/class.ilMailAutoCompleteUserProvider.php:
```diff
@@ -197,7 +197,7 @@ protected function getOrderByPart()
      * @return string
      */
-    protected function getQueryConditionByFieldAndValue($field, $a_str)
+    protected function getQueryConditionByFieldAndValue(string $field, mixed $a_str) : string
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSearchGUI.php:
```diff
@@ -1,66 +1,47 @@
-     * @var ilAccessHandler
-     */
-    protected $access;
+    protected mixed $objParticipants;
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/Form/class.ilIncomingMailInputGUI.php:
```diff
@@ -83,7 +76,7 @@ public function render()
      * @inheritdoc
      */
-    public function getItemByPostVar($a_post_var)
+    public function getItemByPostVar($a_post_var) : mixed
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/Form/class.ilManualPlaceholderInputGUI.php:
```diff
@@ -151,7 +124,7 @@ public function insert($a_tpl)
      * @return string|void
      */
-    public function render($ajax = false)
+    public function render(bool $ajax = false) : mixed
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/class.ilFormatMail.php:
```diff
@@ -31,15 +31,15 @@ public function __construct($a_user_id)
     * @return string
     */
-    public function formatReplyMessage()
+    public function formatReplyMessage() : mixed
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/class.ilFormatMail.php:
```diff
@@ -50,7 +50,7 @@ public function formatReplyMessage()
     * @return string
     */
-    public function formatReplySubject()
+    public function formatReplySubject() : mixed
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/class.ilFormatMail.php:
```diff
@@ -95,7 +95,7 @@ public function formatReplyRecipientsForCC()
     * @return string
     */
-    public function formatReplyRecipient()
+    public function formatReplyRecipient() : mixed
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Mail/classes/class.ilFormatMail.php:
```diff
@@ -111,7 +111,7 @@ public function formatReplyRecipient()
     * @return string
     */
-    public function formatForwardSubject()
+    public function formatForwardSubject() : mixed
```
The `mixed` type is not supported in PHP 7.4, which we'll have to support as well in ILIAS 8.

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSearchGUI.php:
```diff
@@ -1,66 +1,47 @@
-     */
-    protected $mail_roles;
-
+    private RequestInterface $httpRequest;
```
Should be `ServerRequestInterface` IMO

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchCoursesMembersTableGUI.php:
```diff
@@ -16,6 +18,7 @@
 */
 class ilMailSearchCoursesMembersTableGUI extends ilTable2GUI
 {
+    private RequestInterface $httpRequest;
```
Should be `ServerRequestInterface` IMO

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchCoursesTableGUI.php:
```diff
@@ -9,11 +9,15 @@
 
 class ilMailSearchCoursesTableGUI extends ilTable2GUI
 {
+    private RequestInterface $httpRequest;
```
Should be `ServerRequestInterface` IMO

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchGUI.php:
```diff
@@ -29,34 +32,14 @@
-    private $errorDelete = false;
+    private ilGlobalPageTemplate $tpl;
+    private ilCtrl $ctrl;
+    private RequestInterface $httpRequest;
```
Should be `ServerRequestInterface` IMO

mjansenDatabay commented on Services/Contact/classes/class.ilMailingListsGUI.php:
```diff
@@ -14,49 +16,23 @@
-     * @var ilTabsGUI
-     */
-    protected $tabs;
+    private RequestInterface $httpRequest;
```
Should be `ServerRequestInterface` IMO

mjansenDatabay commented on Services/Mail/classes/class.ilAccountMail.php:
```diff
@@ -12,12 +14,13 @@
 */
 class ilAccountMail
 {
+    private RequestInterface $httpRequest;
```
Should be `ServerRequestInterface` IMO

mjansenDatabay commented on Services/Mail/classes/class.ilMailCronNotification.php:
```diff
@@ -11,31 +13,22 @@
-     * @var bool
-     */
-    protected $initDone = false;
+    private RequestInterface $httpRequest;
```
Should be `ServerRequestInterface` IMO

mjansenDatabay commented on Services/Contact/classes/class.ilMailSearchGUI.php:
```diff
@@ -29,34 +32,14 @@
-    private $umail = null;
-
-    private $errorDelete = false;
+    private ilGlobalPageTemplate $tpl;
```
Should be `GlobaTemplateInterface` IMO

mjansenDatabay commented on Services/Contact/classes/class.ilMailingListsGUI.php:
```diff
@@ -14,49 +16,23 @@
-     */
-    protected $tabs;
+    private RequestInterface $httpRequest;
+    protected ilGlobalPageTemplate $tpl;
```
Should be `GlobaTemplateInterface` IMO

mjansenDatabay commented on Services/Mail/classes/class.ilMailAttachmentGUI.php:
```diff
@@ -8,29 +8,14 @@
-
-    /** @var \Psr\Http\Message\ServerRequestInterface */
-    private $request;
+    private ilGlobalPageTemplate $tpl;
```
Should be `GlobaTemplateInterface` IMO

mjansenDatabay commented on Services/Contact/classes/class.ilMailMemberSearchGUI.php:
```diff
@@ -1,66 +1,47 @@
-    protected $access;
+    protected mixed $objParticipants;
+    protected ilCtrl $ctrl;
+    protected ilGlobalPageTemplate $tpl;
```
Should be `GlobaTemplateInterface` IMO

fhelfer committed:
resolving suggested changes

fhelfer committed:
resolve some more

fhelfer committed:
contact classes remove unnecessary include & bugifx mailing lists

fhelfer committed:
removed includes, added static & type to callbacks

fhelfer committed:
cs-fixer remove docs with no value

fhelfer committed:
class names replaced with ::class

fhelfer committed:
bugfixing, removed type casts

fhelfer committed:
array type hints added

fhelfer committed:
bugfixing

mjansenDatabay committed:
Contacts: PHP 8

mjansenDatabay commented:
@fhelfer I added a commit which finalizes `Services/Contacts`.

Please apply the following changes to `Services/Mail`:

- [ ] Please remove all `require_once` and `include_once` directives.
- [ ] Please apply the `PSR 2 + x` code style to the `Services/Mail` folder. There are at least 10 files which violate our coding style. Use `php-cs-fixer`: `libs/composer/vendor/friendsofphp/php-cs-fixer/php-cs-fixer fix --config=./CI/PHP-CS-Fixer/code-format.php_cs --using-cache=no Services/Mail/classes/`
- [ ] There a still some `PhpDoc` issues. Please have a look at my comments **above**. Remove obsolete `PhpDoc` comments even for `__construct`. Don't touch `RFC822.php`.
- [ ] In class `ilMail` you declared many properties as `optional`. I am sure they are not. Please check other classes (like `ilMailOptionsGUI`, ...) as well.
- [ ] Please remove leading `\` in front of classes where the `\` is obsolete.
- [ ] `str_starts_with` is not supported in `PHP 7.4`, please use `strpos($haystack, $needle) === 0` instead. I know that there might be a composer polyfill which make it work in `PHP 7.4`, but we should not use it.
- [ ] In some files the recommended line length is exceeded. Try to use multiple lines instead.
- [ ] In the meantime the `Refinery` can be used together with our `HTTP Service`. Try to use `$DIC->http()->wrapper()->query()` and `$DIC->http()->wrapper()->post()` with the `has` and `retrieve` methods for processing the user input. You'll have to pass a `Refinery Transformation` when retrieving a value. Please have a look at `Services/Contacts` for examples (or contact me).
- [ ] Please remove obsolete blank lines in the classes (between methods and property declarations).
- [ ] Please cast values (e.g. integers with `(int)`) of the array in `\ilMail::fetchMailData`. Otherwise, there will be some problems in the consumer code when working with this array, e.g. the `sender_id` accessed in `\ilPDMailGUI::getPDMailHTML` (and many other places).

fhelfer committed:
removed require_once/include_once, copyright changed, deleted obsolete optional declarations

fhelfer committed:
remove php docs

fhelfer committed:
break exceeding lines

fhelfer committed:
replace http->request() with wrapper & refinery

fhelfer committed:
bugfix fileupload

fhelfer committed:
quick fix

mjansenDatabay committed:
Mail: PHP 8

mjansenDatabay committed:
Mail: PHP 8

fhelfer committed:
removed blank lines, added Select integer casts

mjansenDatabay committed:
Mail: PHP 8

mjansenDatabay committed:
Mail: PHP 8

fhelfer committed:
Merge branch 'php8-mail-migration' of github.com:fhelfer/ILIAS into php8-mail-migration

mjansenDatabay committed:
Mail: PHP 8

mjansenDatabay committed:
Mail: PHP 8

fhelfer committed:
php8 changes Services/Mail/classes

fhelfer committed:
php8 changes Services/Mail/classes

fhelfer committed:
cs fix ilObjMail

fhelfer committed:
cs fix ilObjMail

fhelfer committed:
shorten null check syntax in contructor

fhelfer committed:
shorten null check syntax in contructor

mjansenDatabay committed:
Mail: PHP 8

fhelfer committed:
build artifacts changes

fhelfer committed:
merge resolve

mjansenDatabay committed:
Mail: PHP 8
